### ISSUE TITLE
[otbn] data loading from elf via dmem broken in ISS

### ISSUE THREAD

**Issue by felixmiller** [lowRISC/opentitan#3604]
In current master loading data from elf via dmem is broken somehow. It seems to be an issue already with populating the dmem array from elf data.

I cherry-picked #3589 but it seems to be still there.

Here is a minimal example reproducing the issue:
```
.text
start:
bn.lid x0, 0(x0)
ecall

.data
.quad 0x0000000000000001
.quad 0x0000000000000002
.quad 0x0000000000000003
.quad 0x0000000000000004
```

This is the output on master:

```
bn.lid       x0, 0(x0)              | [otbn.STATUS |= 0x000001 (from HW) (now 0x000001), w0 = 0000000100000000000000020000000000000003000000000000000400000000]
(stall)                             | []
ecall                               | [otbn.INTR_STATE |= 0x000001 (from HW) (now 0x000001), otbn.STATUS &= ~ 0x000001 (from HW) (now 0x000000)]

```
This is the output with #3589 (something weird going on with the leading zeros in the print, but the issue seems to be still there)
```
bn.lid       x0, 0(x0)              | [otbn.STATUS |= 0x000001 (from HW) (now 0x000001), w0 = 0x0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000100000000000000020000000000000003000000000000000400000000]
(stall)                             | []
ecall                               | [otbn.INTR_STATE |= 0x000001 (from HW) (now 0x000001), otbn.STATUS &= ~ 0x000001 (from HW) (now 0x000000)]
```

**Comment by rswarbrick** [comment]
Oh dear! I'm looking in to this now.

**Comment by rswarbrick** [comment]
This is (as always) an endianness issue. The code currently interprets a 256-bit chunk of ELF data as 8 little-endian 32-bit words. Your code assumes we'll interpret it as 4 little-endian 64-bit words.

I'm happy to change to this (especially if that's what we used to do), but it seems a little arbitrary: 64-bit words don't come up in the system in any other context.

As a short-term work-around, I think you can just permute the constants in your file, changing

    .word 0123456789abcdef

to

    .word 89abcdef01234567

so the example above becomes

	.quad 0x0000000100000000
	.quad 0x0000000200000000
	.quad 0x0000000300000000
	.quad 0x0000000400000000

or, maybe less mysteriously,

    .dword 0
    .dword 0x00000001
    .dword 0
    .dword 0x00000002
    .dword 0
    .dword 0x00000003
    .dword 0
    .dword 0x00000004

More generally, we should decide on our "ELF interchange format" :-) I *think* that the current code should match what you'd need if you were loading the data in from Ibex one word at a time. But maybe I've got it wrong?

@imphil, @tomroberts-lowrisc, @GregAC: What do you think?

**Comment by felixmiller** [comment]
I think you're right it, we probably should rather not use `.quad`. I was only using this, since it is the directive taking the largest numbers (fewer lines of code). I guess there is no 256 bit directive, right?

> The code currently interprets a 256-bit chunk of ELF data as 8 little-endian 32-bit words. Your code assumes we'll interpret it as 4 little-endian 64-bit words.

Not sure if I follow, isn't this the same at the end?

My expectation would have been that, if I enumerate each individual byte:

```
.data
.dword 0x03020100
.dword 0x07060504
.dword 0x0B0A0908
.dword 0x0F0E0D0C
.dword 0x13121110
.dword 0x17161514
.dword 0x1B1A1918
.dword 0x1F1E1D1C
```
and
```
.data
.quad 0x0706050403020100
.quad 0x0F0E0D0C0B0A0908
.quad 0x1716151413121110
.quad 0x1F1E1D1C1B1A1918
```
would be equivalent and that both will populate the first dmem cell with `0x1F1E1D1C1B1A191817161514131211100F0E0D0C0B0A09080706050403020100`

but when I do a hexdump with readelf, I get:

for the `.dword` version
```
Hex dump of section '.data':
  0x00000000 00010203 00000000 04050607 00000000 ................
  0x00000010 08090a0b 00000000 0c0d0e0f 00000000 ................
  0x00000020 10111213 00000000 14151617 00000000 ................
  0x00000030 18191a1b 00000000 1c1d1e1f 00000000 ................
```
and for the `.quad` version:

```
Hex dump of section '.data':
  0x00000000 00010203 04050607 08090a0b 0c0d0e0f ................
  0x00000010 10111213 14151617 18191a1b 1c1d1e1f ................
```

**Comment by felixmiller** [comment]
ah, `.dword` is also 64 bit, I thought that would have been 32 bit. So `.word` is 32 bit then. If I do the experiment from above with `.word` I get what I expected for 32 bit constants (same as with `.quad`)
```
Hex dump of section '.data':
  0x00000000 00010203 04050607 08090a0b 0c0d0e0f ................
  0x00000010 10111213 14151617 18191a1b 1c1d1e1f ................
```

**Comment by GregAC** [comment]
> This is (as always) an endianness issue. The code currently interprets a 256-bit chunk of ELF data as 8 little-endian 32-bit words.

This sounds like the correct behaviour to me. OTBN can either load/store 256 bits at a time or 32 bits at a time (aligned in both cases) so those are the only two views we need to worry about.

From @felixmiller's latest comments it sounds like provided you use `.word` in the assembly file the current behaviour works as expected?

Probably worth a note in the assembly style guide that .word is preferred over .quad for this reason.

**Comment by felixmiller** [comment]
@GregAC my confusion regarding `.quad`, `.word` and `.dword` is gone, but the issue itself (if it is one) still persists.

To sumarize, with both of the following versions:

```
.data
.quad 0x0706050403020100
.quad 0x0F0E0D0C0B0A0908
.quad 0x1716151413121110
.quad 0x1F1E1D1C1B1A1918

```
and

```
.data
.word 0x03020100
.word 0x07060504
.word 0x0B0A0908
.word 0x0F0E0D0C
.word 0x13121110
.word 0x17161514
.word 0x1B1A1918
.word 0x1F1E1D1C
```

I get the same hexdump from elf (which seems right for me) which is:

```
  Hex dump of section '.data':
  0x00000000 00010203 04050607 08090a0b 0c0d0e0f ................
  0x00000010 10111213 14151617 18191a1b 1c1d1e1f ................
```
Obviously I then also get the same trace (below). Where for me the 32 bit words seem to be in the wrong order when reading into a wreg. When I load 32 bit words on the other hand I get what I expected:
```
bn.lid       x0, 0(x0)              | [otbn.STATUS |= 0x000001 (from HW) (now 0x000001), w0 = 0x03020100070605040b0a09080f0e0d0c13121110171615141b1a19181f1e1d1c]
(stall)                             | []
lw           x2, 0(x0)              | [x2 = 0x03020100]
(stall)                             | []
lw           x2, 4(x0)              | [x2 = 0x07060504]
(stall)                             | []
lw           x2, 8(x0)              | [x2 = 0x0b0a0908]
(stall)                             | []
lw           x2, 12(x0)             | [x2 = 0x0f0e0d0c]
(stall)                             | []
lw           x2, 16(x0)             | [x2 = 0x13121110]
(stall)                             | []
lw           x2, 20(x0)             | [x2 = 0x17161514]
(stall)                             | []
lw           x2, 24(x0)             | [x2 = 0x1b1a1918]
(stall)                             | []
lw           x2, 28(x0)             | [x2 = 0x1f1e1d1c]
(stall)                             | []
ecall                               | [otbn.INTR_STATE |= 0x000001 (from HW) (now 0x000001), otbn.STATUS &= ~ 0x000001 (from HW) (now 0x000000)]

```

**Comment by rswarbrick** [comment]
@felixmiller: Sorry, I'd messed up with the `.dword` example (and presumably had forgotten to re-build: oops!). If you replace every occurrence of ".dword" with ".word" in my previous message, I think it makes sense.

You're right that the ordering is still a bit confusing. Let's just specify what ordering we want and I'll amend the simulator to match. **We should also write this down in the spec, since it's architecturally visible**

At the moment, we're loading/storing memory "big-endian by 32-bit word and then little-endian within each 32 bits". This does seem a bit odd. Do we just say 256-bit little endian? Then the memory contents you show above will come out as `0x1f1e1d ... 020100`.

**Comment by felixmiller** [comment]
> Do we just say 256-bit little endian?

Yes, that would be preferred from my side. This was also the behavior in the ISS until just recently right? At least, I was already able to compute the correct Montgomery parameters based on the constants I loaded via `.quad` assuming it is 256-bit little endian

**Comment by rswarbrick** [comment]
Cool. Sounds like I broke it when writing `dmem.py`. I'll push up a fix (and spec clarification) in a minute.

**Comment by felixmiller** [comment]
Fixed with #3612
